### PR TITLE
ec2_asg: add AWSRetry to boto3 calls

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Add AWSRetry to boto3 calls in ec2_asg. 2.3 currently has retry capability but the module was ported to boto3 for 2.4 and had to be redone (AWSRetry itself was not an option for 2.3 since it handles boto3 exceptions).

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_asg.py

##### ANSIBLE VERSION
```
2.4.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_asg.py [hunk 1]
```diff
@@ -394,6 +394,104 @@
 
 INSTANCE_ATTRIBUTES = ('instance_id', 'health_status', 'lifecycle_state', 'launch_config_name')
 
+backoff_params = dict(tries=10, delay=3, backoff=1.5)
+
+@AWSRetry.backoff(**backoff_params)
+def describe_autoscaling_groups(connection, group_name):
+    response = connection.describe_auto_scaling_groups(AutoScalingGroupNames=[group_name])
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: This should be paginated: http://boto3.readthedocs.io/en/latest/reference/services/autoscaling.html#paginators

File: lib/ansible/modules/cloud/amazon/ec2_asg.py [hunk 2]
```diff
@@ -394,6 +394,104 @@
...
+    connection.suspend_processes(AutoScalingGroupName=asg_name, ScalingProcesses=processes)
+
+
+@AWSRetry.backoff(**backoff_params)
+def resume_asg_processes(connection, asg_name, processes):
+    connection.resume_processes(AutoScalingGroupName=asg_name, ScalingProcesses=processes)
+
+
+@AWSRetry.backoff(**backoff_params)
+def describe_launch_configurations(connection, launch_config_name):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py:435]: This should be paginated

**Review by willthames — CHANGES_REQUESTED** [review]
This looks pretty good. The pagination is worth doing while we're in the area, and the tests need fixing, but this should be relatively straightforward after that.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py [hunk 3]
```diff
@@ -396,10 +396,24 @@
 
 backoff_params = dict(tries=10, delay=3, backoff=1.5)
 
+
+def paginated_describe_asg(connection, group_name):
+    pg = connection.get_paginator('describe_auto_scaling_groups')
+    for page in pg.paginate(AutoScalingGroupNames=[group_name]):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: is there a reason why you don't do: 

```
return pg.paginate(AutoScalingGroupNames=[group_name]).build_full_result()
```

(other than it being incredibly poorly documented, which is a fair reason)

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: The reason would just be ignorance. Will fix :)

**Review by willthames — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Pull request merged**